csharp/src/Microsoft.ML.OnnxRuntime/NamedOnnxValue.cs

```diff
@@ -218,12 +218,28 @@ out nativeTensor
                                                     ));
 
                     // fill the native tensor, using GetValue(index) from the Tensor<string>
-                    string[] stringsInTensor = new string[tensorValue.Length];
-                    for (int i = 0; i < tensorValue.Length; i++)
+                    var len = tensorValue.Length;
+                    var stringsInTensor = new IntPtr[len];
+                    var pinnedHandles = new GCHandle[len + 1];
+                    pinnedHandles[len] = GCHandle.Alloc(stringsInTensor, GCHandleType.Pinned);
+                    try
                     {
-                        stringsInTensor[i] = tensorValue.GetValue(i);
+                        for (int i = 0; i < tensorValue.Length; i++)
+                        {
+                            var utf8str = UTF8Encoding.UTF8.GetBytes(tensorValue.GetValue(i) + "\0");
+                            pinnedHandles[i] = GCHandle.Alloc(utf8str, GCHandleType.Pinned);
+                            stringsInTensor[i] = pinnedHandles[i].AddrOfPinnedObject();
+                        }
+
+                        NativeApiStatus.VerifySuccess(NativeMethods.OrtFillStringTensor(nativeTensor, stringsInTensor, (UIntPtr)len));
+                    }
+                    finally
+                    {
+                        foreach (var handle in pinnedHandles)
+                        {
+                            handle.Free();
+                        }
```

Comment:
You need to check if the handle is not null since the exception can throw before all of them are created

Reply:
GChandle is a struct object (not class object), so an array of structs should never be null in C#.

However, added a check to see if's been allocated before calling free.